rpm/packagescanner.go

```diff
@@ -161,78 +179,146 @@ func (ps *Scanner) Scan(ctx context.Context, layer *claircore.Layer) ([]*clairco
+		"--no-same-owner",
+		"--no-same-permissions",
+	)
 	tarcmd.Stdin = rd
 	tarcmd.Stderr = &errbuf
+	tarcmd.Dir = root
 	zlog.Debug(ctx).Str("dir", root).Strs("cmd", tarcmd.Args).Msg("tar invocation")
 	if err := tarcmd.Run(); err != nil {
 		zlog.Error(ctx).
 			Str("dir", root).
 			Strs("cmd", tarcmd.Args).
-			Str("stderr", errbuf.String()).
+			Stringer("stderr", &errbuf).
 			AnErr("err", err).
 			Msg("error extracting layer")
+		if rmerr := os.RemoveAll(root); rmerr != nil {
+			err = fmt.Errorf("%v (and during cleanup: %v)", err, rmerr)
+		}
 		return nil, fmt.Errorf("rpm: failed to untar: %w", err)
 	}
 	zlog.Debug(ctx).Str("dir", root).Msg("extracted layer")
-
-	var pkgs []*claircore.Package
-	// Using --root and --dbpath, run rpm query on every suspected database
-	for _, db := range found {
-		zlog.Debug(ctx).Str("db", db).Msg("examining database")
-
-		cmd := exec.CommandContext(ctx, "rpm",
-			`--root`, root, `--dbpath`, db,
-			`--query`, `--all`, `--queryformat`, queryFmt)
-		r, err := cmd.StdoutPipe()
+	if err := filepath.WalkDir(root, func(path string, d os.DirEntry, err error) error {
+		if !d.IsDir() {
+			return nil
+		}
 		if err != nil {
-			return nil, err
+			zlog.Warn(ctx).Err(err).Msg("error walking extracted layer")
 		}
-		errbuf := bytes.Buffer{}
-		cmd.Stderr = &errbuf
-		zlog.Debug(ctx).Str("db", db).Strs("cmd", cmd.Args).Msg("rpm invocation")
-		if err := cmd.Start(); err != nil {
-			r.Close()
-			return nil, err
+		fi, err := d.Info()
+		if err != nil {
+			return err
 		}
-		// Use a closure to defer the Close call.
-		if err := func() error {
-			defer r.Close()
-			srcs := make(map[string]*claircore.Package)
-			s := bufio.NewScanner(r)
-			s.Split(querySplit)
-
-			for s.Scan() {
-				p, err := parsePackage(ctx, srcs, bytes.NewBuffer(s.Bytes()))
-				if err != nil {
-					return err
-				}
-				p.PackageDB = db
-				pkgs = append(pkgs, p)
-			}
+		return os.Chmod(path, fi.Mode()|0700)
+	}); err != nil {
+		if rmerr := os.RemoveAll(root); rmerr != nil {
+			err = fmt.Errorf("%v (and during cleanup: %v)", err, rmerr)
+		}
+		return nil, fmt.Errorf("rpm: failed to change permssions: %w", err)
+	}
 
-			return s.Err()
-		}(); err != nil {
-			if errbuf.Len() != 0 {
-				zlog.Warn(ctx).
-					Str("db", db).
-					Strs("cmd", cmd.Args).
-					Str("err", errbuf.String()).
-					Msg("error output")
-			}
-			return nil, fmt.Errorf("rpm: error reading rpm output: %w", err)
+	l := &extractedLayer{root: root}
+	// This takes the extreme measure of panicing the program if the layer
+	// directory is leaked or unsuccessfully cleaned up.
+	//
+	// Since these can use a nontrivial amount of disk space (layers around 1GiB
+	// are not uncommon) the eventual result is the same, but this way operators
+	// are more likely to notice.
+	_, file, line, _ := runtime.Caller(1)
+	runtime.SetFinalizer(l, func(l *extractedLayer) {
+		panic(fmt.Sprintf("%s:%d: layer directory %q not closed", file, line, root))
```

Comment:
This makes me a little nervous, I understand it's an issue that's probably going to affect the normal running of the application and probably lead to a full disk, but I'm thinking that should be discerned via an errors in the logs. We are sort of assuming an infrastructure by panicking, like we're assuming some kind of orchestration / storage setup.

I think the instance will error out until it's manually restarted with clean storage when the disk is full, but it could service other requests that aren't novel indexing requests. (which I'm not sure is better or worse from a diagnostic POV).